### ISSUE TITLE
Containerize Script Exporter

### ISSUE THREAD

**Issue by xi-yang** [esnet/sense-rtmon#58]
This was a work in progress but @sunami09 @PannuMuthu encounter some in-container shell  environment issues the kept the script from starting. 

@abessiari please help resolve the issue.

**Comment by abessiari** [comment]
@sunami09 @PannuMuthu 

Can you give me the Dockerfile in question. I will try to build the container .... and see what is happening
Thanks.

**Comment by sunami09** [comment]
@abessiari 
This is Prometheus targets status:
![image](https://github.com/esnet/sense-rtmon/assets/66564001/9dc35252-529c-4ef4-9594-a8c7f7c17de9)

The single is down.

cloud/prometheus.yml
```yml
  - job_name: single
    metrics_path: /probe
    params:
      script: [default]
    static_configs:
      - targets:
        - 52.20.53.21:9469 
```

cloud/script_exporter/examples/config.yml

```yml
tls:
  enabled: false
  crt: server.crt
  key: server.key

basicAuth:
  enabled: false
  username: admin
  password: admin

bearerAuth:
  enabled: false
  signingKey: my_secret_key

discovery:
  # host: localhost
  # port: 1231
  # scheme: https

scripts:
  - name: default
    script: ./examples/l2debugging.sh
```
```bash
[root@ip-172-31-72-189 examples]# pwd
/home/centos/sense-rtmon/cloud/script_exporter/examples
[root@ip-172-31-72-189 examples]# ls
archive  args.sh  config.yaml  docker.sh  generate_script.py  helloworld.sh  l2debugging.sh  ping.sh  README.md  showtimeout.sh  test.sh
```

```bash
[root@ip-172-31-72-189 cloud]# docker ps
CONTAINER ID   IMAGE                               COMMAND                  CREATED         STATUS         PORTS      NAMES
1db6eef9cd37   grafana/grafana-enterprise:latest   "/run.sh"                3 minutes ago   Up 3 minutes   3000/tcp   cloud_grafana.1.3lwyr97bllz9tk6of5plgumx6
1202a7092517   ricoberger/script_exporter:v2.5.2   "/bin/script_exporte…"   3 minutes ago   Up 3 minutes   9469/tcp   cloud_script_exporter.1.6aiiff50kq2w5nkatxtbfprhp
3166a268b65a   prom/prometheus:v2.2.1              "/bin/prometheus --c…"   3 minutes ago   Up 3 minutes   9090/tcp   cloud_prometheus.1.p0rnahd40dxe16omv0q09tbg3
b9b4632023d4   prom/pushgateway:latest             "/bin/pushgateway"       3 minutes ago   Up 3 minutes   9091/tcp   cloud_pushgateway.1.wyw9ljnr4f9w3770ajcqq7wlb
[root@ip-172-31-72-189 cloud]# docker logs 1202a7092517
Starting server (version=v2.5.2, branch=HEAD, revision=c4cc5d8651da815b6518694d37187c979f0c897e)
Build context (go=go1.18.1, user=root, date=20220430-11:49:40)
script_exporter listening on :9469
[root@ip-172-31-72-189 cloud]# sudo lsof -i -P -n | grep 9469
dockerd    3821       root   68u  IPv6 1128949667      0t0  TCP *:9469 (LISTEN)
```

**Comment by abessiari** [comment]
@sunami09 @PannuMuthu 

So I get"Script not found" with image `ricoberger/script_exporter:v2.5.2`

I used the latest version `ricoberger/script_exporter:v2.16.0 ` and it is fine 

I tested using the ping script and the the sleep script like so:

http://localhost:9469/probe?script=ping&script=sleep

The output is:

```
# HELP script_success Script exit status (0 = error, 1 = success).
# TYPE script_success gauge
script_success{script="ping"} 1
# HELP script_duration_seconds Script execution time, in seconds.
# TYPE script_duration_seconds gauge
script_duration_seconds{script="ping"} 2.047687
# HELP script_exit_code The exit code of the script.
# TYPE script_exit_code gauge
script_exit_code{script="ping"} 0

# HELP script_success Script exit status (0 = error, 1 = success).
# TYPE script_success gauge
script_success{script="sleep"} 1
# HELP script_duration_seconds Script execution time, in seconds.
# TYPE script_duration_seconds gauge
script_duration_seconds{script="sleep"} 1.006126
# HELP script_exit_code The exit code of the script.
# TYPE script_exit_code gauge
script_exit_code{script="sleep"} 0

```

**Comment by PannuMuthu** [comment]
@abessiari This was the issue we were seeing as well. Here are some solutions we tried to navigate around this issue:

---

### **Verify that the script is copied over to the Docker container filesystem**

We verified that when mounting our bare metal `examples` directory to the Docker container directory, we could see the copied `examples` directory as well as any scripts we manually added to our `examples` directory in bare metal: 

```
docker exec -it <docker_id> /bin/bash
```

### **Verify that the bash scripts are executables**

We verified that the bash scripts were read-write-edit accessible to all users (i.e. `chmod 777`) from within the Docker container filesystem

### **Revert Script Exporter Version to Prior Stable Version**

We reverted script exporter to a prior stable release (Version 2.5.2) where we had seen successful execution and metrics export of `l2debugging.sh` script to Pushgateway.

### **Verify Script Exporter Configuration as Prometheus Target**

We verified that the script exporter metrics page and port (default 9469) were accessible to Pushgateway and the Pushgateway API could successfully query the metrics page via the Prometheus Target Status page: 

![targets](https://github.com/esnet/sense-rtmon/assets/30813477/34a42cee-c9dc-488c-9363-7815335f3bd7)

### **Verify the config file pointed to the correct script filepath**

We verified that the `config.yaml` file within `examples` pointed to the correct filepath for the associated bash script and scripts + config file were in the same directory hierarchy.

---

You are able to reproduce the error we were blocked by (`script not found`). We ended up bypassing this issue by manually running `l2debugging.sh` at our specified scrape interval (every 15s) and pushing these metrics to Pushgateway via `curl` through the main loop in `main.py`.

**Comment by PannuMuthu** [comment]
Thanks for the suggestion of upgrading the script exporter container from 2.5.2 to the latest version (2.16.0)! We have tried to implement this within our workflow, but we encounter a similar error that we have seen in prior attempts but is unreproducible:

We use this command within `docker-stack.yaml` to build the container:

```
  script_exporter:
    command:
      - '-config.file=/examples/config.yaml'
      - '-web.listen-address=:9469'
    image: 'ricoberger/script_exporter:v2.16.0'
    ports:
      - '9469:9469'
    volumes:
      - './script_exporter/examples:/examples'
    deploy:
      placement:
        constraints:
          - node.role==manager
      restart_policy:
        condition: on-failure
    networks:
      - monitor-net 
```

This is analogous to a command such as `docker run --rm -it --name script_exporter -p 9469:9469 -v ./script_exporter/examples:/examples ricoberger/script_exporter:v2.16.0 -config.file /examples/config.yaml
`

**Execution Result of `docker ps`**:

![dockerPS](https://github.com/esnet/sense-rtmon/assets/30813477/d4639283-5758-41f4-96e7-8d719aa0c49e)


**Execution Result of Inspecting Container Filesystem**:

![dockerFS](https://github.com/esnet/sense-rtmon/assets/30813477/16b9bed7-59ba-45ee-96e4-db0f34883e07)

**Execution Result of Inspecting Docker Logs**:

![logs](https://github.com/esnet/sense-rtmon/assets/30813477/eec5d878-f955-4c7d-b7e3-60f1fc395405)

From these inspections, everything seems to look good. However, looking at Prometheus targets or navigating to the Script Exporter metrics page, we find:

![promTargets](https://github.com/esnet/sense-rtmon/assets/30813477/8c16ccf7-aad3-4ded-9a28-eff10cc41770)

**Comment by abessiari** [comment]
@PannuMuthu @sunami09 

Can you try http://localhost:9469/probe?script=default using curl or browser?

I see script="default"  on the single Prometheus target.  I suspect the quotes. It is possible prometheus is encoding the quotes to %22. My brave browser does encode the quotes. Curl command does not ....

http://localhost:9469/probe?script=default
http://localhost:9469/probe?script="default"
http://localhost:9469/probe?script=%22default%22

In any case remove the quotes from the target. They are not needed ....

**Comment by abessiari** [comment]
@PannuMuthu I have invited you to a zoom meeting on slack.
If you can't make it let's have one at 10:00 Pacific time.

**Comment by abessiari** [comment]
@PannuMuthu @sunami09 

Ok I think prometheus is unable to reach the script exporter cause 
its config uses **`52.20.53.21:9469`**.  There is no need to use that 
ip address at all. **Also there is no need to expose port 9469.** The two containers
can talk to each other using hostnames (or vips).... They are in the same network. 

![prometheus](https://github.com/esnet/sense-rtmon/assets/48293960/5f45a634-b66c-4be6-b86e-3d6cc735eebc)

 **Note** `script_exporter:9469` in the prometheus config file and `hostname: script_exporter` in the docker compose file. 

# prometheus config file
```
  - job_name: single
    metrics_path: /probe
    params:
      script: ["sleep"]
    static_configs:
      - targets:
        - **script_exporter:9469**
  ```

# docker-compose file. 
# I launch it like so: `docker stack deploy --compose-file docker-compose.yaml mytest`

```
services:
  script_exporter:
    **hostname: script_exporter**
    container_name: script_exporter
    image: ricoberger/script_exporter:v2.16.0
    volumes:
      - .../examples:/examples
    command:
      - '-config.file=/examples/config.yaml'
  prometheus:
    hostname: prometheus
    container_name: prometheus
    image: prom/prometheus:v2.2.1
    ports:
      - "9090:9090"
    volumes:
      - .../prometheus.yml:/etc/prometheus/prometheus.yml
```

**Comment by abessiari** [comment]
@PannuMuthu 

Any progress on this? I am adding my notes that I sent on slack. 

```
git clone https://github.com/abessiari/script_exporter.git -b test_with_prometheus
cd script_exporter
docker stack deploy --compose-file ./docker-compose.yaml mytest
# Go to http://localhost:9090/targets. Wait a while all targets should be up
```

**Comment by PannuMuthu** [comment]
Thanks @abessiari for your help.

We have resolved the Script Exporter containerization issue and can see metrics at `:9469\metrics` when deploying the container via Docker Stack.

However, while developing the containerization package for the main loop, we realized that it would be simpler to include and call the debugging script within the main loop, as the debugging script must be modified each time a new flow configuration is detected (i.e. a new flow is either `CREATE - READY` or `REINSTATE - READY` status on the SENSE Orchestrator), and this would require a removal and restart of a Script Exporter container if we were to containerize the script. Instead, we are modifying the bash script within the main loop container each time a new flow is detected.
